[ACP-99] ValidatorSetManager Solidity Contract (#99)
avalanche-foundation/ACPs: ACPs/99-validatorsetmanager-contract/README.md

```text
ACP: 99
Title: ValidatorSetManager Solidity Contract
Author(s): Gauthier Leonard <https://github.com/Nuttymoon>
Discussions-To: https://github.com/avalanche-foundation/ACPs/discussions/98
Status: Proposed
Track: Best Practices
Dependencies: 77
```

## Abstract

Define (i) a reference implementation for a minimal Solidity smart contract to manage a Subnet’s validator set from an Avalanche EVM chain, as well as (ii) a reference architecture to easily plug in custom “security modules” on top of this contract (e.g. to implement a PoS Subnet).

The proposed name for this contract is `ValidatorSetManager`.

This proposal should also help define the payload of the P-Chain notifications sent to the Subnet upon validator addition, weight change and removal.

This ACP relies on concepts introduced in [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on it to be marked as `Implementable`.

## Motivation

> [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets) opens the door to managing a Subnet validator set (stored on the P-Chain) from any chain on the Avalanche Network. The P-Chain allows a Subnet to specify a Warp Derived Address as the validator set manager. The `(blockchainID, address)` pair specified by the Warp Derived Address is responsible for sending Warp `AddressedCall` payloads to issue `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` on the P-Chain. This enables an onchain program add, modify the weight of, and remove validators.

Reviewer: In ACP77, don't think there will be a `SetSubnetValidatorWeightTx`, instead this would be the specified tuple (chainID, address) in `ConvertSubnetTx`

Author: https://github.com/avalanche-foundation/ACPs/blob/11e8880a839c4b2020c7c84423c6c9042092487f/ACPs/77-reinventing-subnets/README.md?plain=1#L205

Reviewer: my bad commented on the wrong line, meant `SetSubnetValidatorManagerTx` [here](https://github.com/avalanche-foundation/ACPs/pull/99/files#diff-a7a0317424a0341b7a5426c1c751b11b8683adf4e990c0d1f37c45641459d898R36) is not defined in ACP77, and will instead be set in `ConvertSubnetTx`.

2. Send validator set updates to the P-Chain by generating `AdressedCall`s defined in ACP-77
3. Correctly update the validator set by interpreting notification messages received from the P-Chain
4. Be easily extended with custom “security modules” to implement any security model

Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching Subnets on Avalanche.

Once deployed, this contract will be used as the `address` in the `SetSubnetValidatorManagerTx`.

## Specification

### Storage

The `ValidatorSetManager` needs to store all this data to operate correctly.

**Note:** This is pseudo-code that might not be 100% syntactically correct.

```solidity
// The ID of the Subnet that uses this contract to manage its validator set
bytes32 subnetID;

// The current Subnet validator set (list of NodeIDs)
bytes32[] subnetCurrentValidatorSet;

// Subnet validator
struct SubnetValidator {
    // The Warp message ID corresponding to the RegisterSubnetValidatorTx that added the validator.
    // The message ID is needed to submit subsequent SetSubnetValidatorWeightTx.

>     bytes32 messageID,

Reviewer: think this ID should be a validation ID instead, denoting a hash of the info for a validation period when registering

Author: Absolutely. I am working on an updated version of this ACP reflecting the latest changes in ACP-77!

    // The nonce of the latest validator weight update
    uint64 nonce,
    // The validator weight
    uint64 weight,
    // The timestamp the validator was added
    uint64 addTimestamp
}

// The Subnet current validators
// NodeID => SubnetValidator
mapping(bytes32 => SubnetValidator) subnetCurrentValidators;

// The total weight of the current Subnet validator set
uint64 subnetTotalWeight;

// Subnet validator pending registration
struct SubnetRegisteringValidator {
    // The validator weight
    uint64 weight,
    // The timestamp after which the signature is invalid
    uint64 expiry,
    // The Ed25519 signature over the concatenated bytes of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]

>     bytes signature

Reviewer: ACP77 will use the blsPublicKey only here instead of signature

}

// The Subnet validators that are pending registration (= waiting for addition notification from the P-Chain)
// NodeID => registering validator
mapping(bytes32 => SubnetRegisteringValidator) subnetRegisteringValidators;

// Subnet validation status
enum SubnetValidationStatus {
    Registering,
    Active,
    Removing,
    Over
}

// Subnet validation

> struct SubnetValidation {

Reviewer: Didn't quite get this struct, why do we need to keep a struct for a list of validation periods? Not sure we need a historical list of validation period updates

// Returns a Subnet validator information by its NodeID
function getSubnetValidator(bytes32 nodeID) external view returns (SubnetValidator);

/**
 * Used to issue a RegisterSubnetValidatorTx
 * @param nodeID The ID of the node to add to the Subnet
 * @param weight The weight of the node on the Subnet
 * @param expiry The time after which this message is invalid
 * @param signature The Ed25519 signature of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[timestamp]
 */
function registerSubnetValidator(
    bytes32 nodeID,
    uint64 weight,
    uint64 expiry,
    bytes signature
) external onlyOwner;

/**
 * Used to issue a SetSubnetValidatorWeightTx
 * @param nodeID The ID of the node to modify
 * @param weight The new weight of the node on the Subnet
 */
function setSubnetValidatorWeight(
    bytes32 nodeID,
    uint64 weight
) external onlyOwner;

/**
 * Called to update the Subnet validator set upon receiving a notification Warp message from the P-Chain
 * (after the successful addition or removal of a validator)
 * @param messageIndex The index of the Warp message from the P-Chain
 */
function updateSubnetValidatorSet(
    uint32 messageIndex

> ) external;

Author: ```suggestion
) external;

/** Used to issue a SetSubnetValidatorManagerTx
  * @param blockchainID The ID of the chain on which the Subnet validator manager is located
  * @param managerAddress The address of the Subnet validator set manager
  */
function setSubnetValidatorManager(
    bytes32 blockchainID,
    address managerAddress
) external onlyOwner;
```
This function is gonna be needed in case we want to change the validator manager address of the Subnet.

Author: `SetSubnetValidatorManagerTx` was removed from ACP-77

```

#### About `ExitValidatorSetTx`

This transaction allows a validator to exit a Subnet validator set directly from the P-Chain to prevent the Subnet from locking validators into participating in consensus indefinitely. Therefore it is not meant to be called in the `ValidatorSetManager` contract.

### Information Needed from the P-Chain

> Upon successful addition, weight update, and removal of a validator from the Subnet, the P-Chain will send a notification in the form of a Warp message. The `updateSubnetValidatorSet` function is responsible for deserializing this message and updating the contract state accordingly.

Reviewer: To be more clear here, the P-Chain is willing to sign a Warp message corresponding to validator set updates, but there still needs to be a delivery of that P-Chain message to the ValidatorManager to make it aware, and the delivery of the message is not part of the P-Chain's responsibility.

|     nonce :   uint64 |   8 bytes |
+-----------+----------+-----------+
|    weight :   uint64 |   8 bytes |
+-----------+----------+-----------+
| timestamp :   uint64 |   8 bytes |
+-----------+----------+-----------+
|    uptime : [64]byte |  64 bytes |
+-----------+----------+-----------+
                       | 184 bytes |
                       +-----------+
```

- The Subnet ID
- The NodeID
- The `MessageID` corresponding to the Warp message in the `RegisterSubnetValidatorTx` that added the validator (SHA256)
- The `nonce` corresponding to this update
- The weight of the validator (if `0`, this means that the validation is over)
- The exact timestamp at which the weight was updated (if the nonce is `0`, this also means the beginning of the `validation`)
- The uptime observed during the previous validation period (only useful in case a validator weight update)

Note that the timestamp can also be used to update the `endTime` of the previous validation period.

**Note:** Once this discussion/ACP is in an advanced stage, an update to ACP-77 can be made to specify the payload of P-Chain notifications.

### Reference Architecture for Custom “Security Modules”

> The `ValidatorSetManager` contract will be ownable (e.g. using [OpenZeppelin Ownable](https://docs.openzeppelin.com/contracts/5.x/api/access#Ownable)). The ownership can be transferred to a smart contract, the “Security Module”, that will implement the security rules for the Subnet.

Reviewer 2: Given how crucial this module will be, it may be best practice to enforce a two-step mechanism

Author: Could you expand on what those 2 steps would be?

Reviewer 3: OZ's [Ownable2Step](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable2Step.sol) for example. the new owner needs to accept the ownership transfer before it becomes effective. though this might be harder to handle if the owner is not a smart account but a regular contract.

Reviewer: The likely case for using a contract instead of smart account would be PoS validator sets, in which case the contract interface should likely have an `acceptOwnership` function, and should already be taking in the `ValidatorManager` address. 

Mostly a question of not introducing extra access controls for making `acceptOwnership` call, thinking if this would be done in the constructor

Other “Security Modules” could implement PoS, Liquid PoS, etc. The specification of such smart contracts is out of the scope of this ACP.

## Backwards Compatibility

The `ValidatorSetManager` is only a reference implementation, it doesn’t have any impact on the current behavior of the protocol.

## Reference Implementation

A full reference implementation has not been provided yet. It will be provided once this ACP is considered `Implementable`.

## Security Considerations

The audit process of the `ValidatorSetManager` contract is of the utmost importance for the future of the Avalanche ecosystem as most Subnets would rely upon it to secure their Subnet.

## Open Questions

### Should the Subnet validators’ uptime be tracked on the P-Chain?

Given the direction taken by the Avalanche protocol by the introduction of SOVs (Subnet Only Validators), is it the job of the P-Chain to track Subnet validators’ uptime? If yes, the induced load should be taken into account to compute the continuous fee discussed in ACP-77.

### How could we name “Security Modules”?

> I don’t really like this name but cannot come up with anything else.

Reviewer 2: Maybe the term "Gating" fits better? "Validator Gating Modules"

```text
ACP: 99

> Title: ValidatorSetManager Solidity Contract

Author: This ACP isn't limited to "staking" (e.g. PoA could be implemented on top of the `ValidatorSetManager`)

Author(s): Gauthier Leonard <https://github.com/Nuttymoon>
Discussions-To: https://github.com/avalanche-foundation/ACPs/discussions/98
Status: Proposed
Track: Best Practices
```

## Abstract

Define (i) a reference implementation for a minimal Solidity smart contract to manage a Subnet’s validator set from an Avalanche EVM chain, as well as (ii) a reference architecture to easily plug in custom “security modules” on top of this contract (e.g. to implement a PoS Subnet).

The proposed name for this contract is `ValidatorSetManager`.

This proposal should also help define the payload of the P-Chain notifications sent to the Subnet upon validator addition and removal.

> This ACP relies on concepts introduced in [ACP-73 (Warp Addressed Transactions)](https://github.com/avalanche-foundation/ACPs/pull/73) and [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on them to transition to an `Implementable` state.

Author: I'll look into that.

> This ACP relies on concepts introduced in [ACP-73 (Warp Addressed Transactions)](https://github.com/avalanche-foundation/ACPs/pull/73) and [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on them to transition to an `Implementable` state.

Author: Added a `Dependencies` field in https://github.com/avalanche-foundation/ACPs/pull/99/commits/24d6f561fd4de8d25d7fa725abe55f541a749a43

Not sure this is allowed tho haha. Btw we could add this to ACP-83 while the PR is still open.

## Motivation

[ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets) opens the door to managing a Subnet validator set (stored on the P-Chain) from another chain of the Avalanche network. This is achieved by sending updates to the P-Chain from a “validator manager” address on this chain. The “validator manager” address will be the only one able to issue `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` on the P-Chain to add, change the weight of, and remove validators.

> `Warp Addressed Transactions` (introduced in [ACP-73](https://github.com/avalanche-foundation/ACPs/pull/73)) could be used to send the validator set updates to the P-Chain.

Reviewer 4: I think we should remove mention of ACP-73 since it won't be merged into the ACP repo (cc @aaronbuchwald to confirm)

Reviewer 5: We talked about this and I said it was fine to include to show the progression like including a link to an article 🤷

Reviewer 5: I talked with @Nuttymoon about this and I said it was fine to include to show the progression of ideas like referencing an article 🤷

Reviewer 4: The mention of ACP-73 in the "Abstract" makes sense to me for that reason. This one makes it seem like this ACP relies on ACP-73 being activated with the "Warp Addressed Transactions" (I believe you pointed this out too here https://github.com/avalanche-foundation/ACPs/pull/99#discussion_r1628354647)

Author: I guess it is fine to remove ACP-73 references altogether as it seems that using the `AddessedCall` format is enough to issue transactions on the P-Chain.

Upon addition or removal of a validator, the P-Chain is expected to send a Warp message back to the Subnet. The Subnet needs a way to interpret this message and trigger appropriate actions. These notification messages have not been specified to date.

> Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based (especially the C-Chain, on which we can expect most of the Subnet “validator manager” addresses will be hosted for security reasons), we propose defining a reference implementation for a Solidity smart contract that can:

Author: Do you reckon I should remove this remark?

> Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based (especially the C-Chain, on which we can expect most of the Subnet “validator manager” addresses will be hosted for security reasons), we propose defining a reference implementation for a Solidity smart contract that can:

Reviewer 5: I would since I don't personally expect this to be the case and don't think it's necessary for this ACP

> 1. Hold relevant information about the current Subnet validator set, as well as historical information

Author: If we place ourselves in the context of a PoS xL1, when a validator leaves the validator set we should be able to compute staking rewards based on the historical evolution of its weight and uptime. It is easier to do if all this information is directly queryable on-chain, rather than having to do an asynchronous query to the P-Chain. This is the whole purpose of the `SubnetValidationPeriod` struct defined below.

That being said, tracking the total validator set weight of the xL1 over time would be pretty cumbersome, since it changes at every single validator weight change.

> 2. Send validator set updates by issuing transactions defined in ACP-77 to the P-Chain

Author: ```suggestion
2. Send validator set updates to the P-Chain by generating `AdressedCall`s defined in ACP-77
```

3. Correctly update the validator set by interpreting notification messages received from the P-Chain
4. Be easily extended with custom “security modules” to implement any security model

Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching Subnets on Avalanche.

Once deployed, this contract will be used as the `address` in the `SetSubnetValidatorManagerTx`.

## Specification

### Storage

The `ValidatorSetManager` needs to store all this data to operate correctly.

**Note:** This is pseudo-code that might not be 100% syntactically correct.

```solidity
// The ID of the Subnet that uses this contract to manage its validator set
bytes32 subnetID;

// The current Subnet validator set (list of NodeIDs)

> bytes32[] subnetCurrentValidatorSet;

Reviewer 4: It will be 32 bytes since Subnet Validators will exclusively use Ed25519 public keys

> bytes32[] subnetCurrentValidatorSet;

Author: Yes, I took ACP-77 as reference where the `nodeID` is 32 bytes.

> bytes32[] subnetCurrentValidatorSet;

Author: Actually, I am a bit confused here. NodeIDs will always be 20 bytes no? Even for ACP-77.

> bytes32[] subnetCurrentValidatorSet;

Reviewer 4: No it will be 32 bytes to house the Ed25519 public key. ACP-20 includes details on how this will be done: https://github.com/avalanche-foundation/ACPs/blob/218b62f4dd907f80f5991455dddcf72d59f33bb6/ACPs/20-ed25519-p2p/README.md

> bytes32[] subnetCurrentValidatorSet;

Author: Indeed I totally forgot about this ACP. Any timeline for implementation and rollout?

// Subnet validation status
enum SubnetValidationStatus {
    Registering,
    Active,
    Removing,
    Over
}

// Subnet validation
struct SubnetValidation {
    // The validation status
    SubnetValidationStatus status,
    // The list of validation periods
    // The index is the nonce associated with the weight update.
    SubnetValidationPeriod[] periods,
    // The nonce of the current validation period
    uint64 currentValidationPeriod
}

// Subnet validation period
struct SubnetValidationPeriod {
    // The weight of the validator during the period
    uint64 weight,
    // The start time of the validation period
    uint64 startTime,
    // The end time of the validation period (only ≠ 0 when the period is over)
    uint64 endTime,
    // The uptime of the validator during this validation period (only ≠ 0 once the validation period is over)
    uint64 uptime
}

/**
 * The list of message IDs associated with a validator of the Subnet
 * Each message ID corresponds to a RegisterSubnetValidatorTx
 * NodeID => messageID[]
 */
mapping(bytes32 => bytes32[]) subnetValidatorMessages;
// The validation corresponding to each message ID (= RegisterSubnetValidatorTx)

> // messageID => SubnetValidation
> mapping(bytes32 => SubnetValidation) subnetValidations;

Author: ```suggestion
// NodeID => messageID => SubnetValidation
mapping(bytes32 => mapping(bytes32 => SubnetValidation) subnetValidations;
```
Here it might be more relevant to add a layer with the `NodeID` to avoid having to query `subnetValidatorMessages` beforehand if we already have the `messageID`

## Abstract

Define (i) a reference interface for a minimal Subnet Manager Solidity smart contract to be deployed on any Avalanche EVM chain, as well as (ii) a modular architecture to easily plug in custom “security modules” on top of this contract (e.g. to implement a PoS Subnet).

This ACP relies on concepts introduced in [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on it to be marked as `Implementable`.

## Motivation

[ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets) opens the door to managing a Subnet validator set (stored on the P-Chain) from any chain on the Avalanche Network. The P-Chain allows a Subnet to specify a "Subnet manager". The `(blockchainID, address)` pair specified by the Warp Derived Address is responsible for sending Warp `AddressedCall` payloads to issue `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` on the P-Chain. This enables an onchain program to add, modify the weight of, and remove validators.

> On each validator set change, the P-Chain is willing to sign an `AddressedCall` to notify any onchain program tracking the validator set. Onchain programs must be able to interpret this message, so they can trigger the appropriate action. The 2 kinds of `AddressedCall`s [defined in ACP-77](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#proof-of-subnet-validator-set-change) are `SubnetValidatorRegistrationMessage` and `SubnetValidatorWeightUpdateMessage`.

Author: Created a related comment on ACP-77 PR haha

https://github.com/avalanche-foundation/ACPs/pull/149/files#r1790345739

        Expired
    }

    /**
     * @notice Subnet validation
     * @param status The validation status
     * @param nodeID The NodeID of the validator
     * @param startTime The start time of the validation
     * @param endTime The end time of the validation
     * @param periods The list of validation periods.
     * The index is the nonce associated with the weight update.
     * @param activeSeconds The time during which the validator was active during this validation
     * @param uptimeSeconds The uptime of the validator for this validation
     */
    struct Validation {
        ValidationStatus status;
        bytes32 nodeID;
        uint64 startTime;
        uint64 endTime;
        ValidationPeriod[] periods;
        uint64 activeSeconds;
        uint64 uptimeSeconds;
    }

    /**
     * @notice Subnet validation period
     * @param weight The weight of the validator during the period
     * @param startTime The start time of the validation period
     * @param endTime The end time of the validation period (only ≠ 0 when the period is over)
     */
    struct ValidationPeriod {
        uint64 weight;
        uint64 startTime;
        uint64 endTime;

>     }

Author: - Yes, it is possible to use `validation.periods[0].startTime` instead of `validation.startTime` to avoid duplicating data. For `endTime` it would be a little bit trickier: `validation.periods[validation.periods.length - 1].endTime` (but irrelevant to store `endTime` anyway if we choose to remove history
- I thought it was enough to use the array index as the place to store the nonce since it is incremented by 1 at each weight change
- I chose to track `uptimeSeconds` at the `Validation` level because I don't think it is useful to track it at the `ValidationPeriod` level and thought it might be difficult to differentiate the uptime from a specific period than from the whole validation. The latter depends on how the uptime is tracked at the VM level. Do we have more information about how this will be handled btw?

About the validation historic tracking, I agree that it is most likely not needed. Thus my open question at the end.

    /**
     * @notice Initiate a validator weight update by issuing a SetSubnetValidatorWeightTx Warp message.
     * If the weight is 0, this initiates the removal of the validator from the Subnet. An uptime proof can be
     * included. This proof might be required to claim validator rewards (handled by the security module).
     * @param nodeID The ID of the node to modify
     * @param weight The new weight of the node on the Subnet
     * @param includesUptimeProof Whether the uptime proof is included in the message
     * @param messageIndex The index of the Warp message containing the uptime proof
     */
    function initiateValidatorWeightUpdate(
        bytes32 nodeID,
        uint64 weight,
        bool includesUptimeProof,
        uint32 messageIndex
    ) external;

    /**
     * @notice Completes the validator weight update process by returning an acknowledgement of the weight update of a
     * validationID from the P-Chain.
     * @param messageIndex The index of the Warp message to be received providing the acknowledgement.
     */
    function completeValidatorWeightUpdate(uint32 messageIndex) external;
}
```

#### About `Validation`s

> A `Validation` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator weight changes during the same validation.

Reviewer 6: Tracking independent periods each time the validator's weight changes during the same validation is interesting.

I find it more intuitive to consider a validator's own (initial) weight as static for the entire validation period, and consider any changes to the weight as delegations. I think that makes it easier for the security module (or whichever contract would be responsible for potential rewards) to compute the rewards for a given validator or delegator.

Author: It is indeed more complicated.

What I had in mind was the auto-compounding workflow for validators. Ending a validation will most likely trigger a delay period during which tokens will be idle. Offering the capacity to change the weight during a validation allows us to circumvent this problem.

In my implementation, I compute the average of the validator weight during the whole validation period in `_notifySecurityModuleValidatorWeightUpdate`. Happy to have your thoughts: https://github.com/suzaku-network/suzaku-contracts-library/blob/1d13efa366ddd4b288a884351402c3492f65e15d/contracts/src/contracts/ACP99/ACP99Manager.sol#L395

     * @param nonce A sequential number to order weight changes
     * @param newWeight The new weight assigned to the validator
     * @param uptime The uptime information for the validator
     */
    struct ValidatorWeightChangeInfo {
        bytes32 nodeID;
        bytes32 validationID;
        uint64 nonce;
        uint64 newWeight;
        ValidatorUptimeInfo uptimeInfo;
    }

    error ACP99SecurityModule__ZeroAddressManager();
    error ACP99SecurityModule__OnlyManager(address sender, address manager);

    /// @notice Get the address of the ACP99Manager contract secured by this module
    function getManagerAddress() external view returns (address);

    /**
     * @notice Handle a validator registration
     * @param validatorInfo The information about the validator
     */
    function handleValidatorRegistration(ValidatiorRegistrationInfo memory validatorInfo)
        external;

    /**
     * @notice Handle a validator weight change
     * @param weightChangeInfo The information about the validator weight change
     */
    function handleValidatorWeightChange(ValidatorWeightChangeInfo memory weightChangeInfo)
        external;
}
```

### Reference Architecture

> Each `ACP99Manager` contract will be associated with one "security module" that must implement the `IACP99SecurityModule` interface and is the only contract allowed to call the `ACP99Manager` functions related to validator set changes (`initiateValidatorRegistration`, and `initiateValidatorWeightUpdate`). Everytime a validator is added/removed or a weight change is initiated, the `ACP99Manager` will in turn call the corresponding function of the "security module" (`handleValidatorRegistration` or `handleValidatorWeightChange`). We recommand that the "security module" reference an immutable `ACP99Manager` contract address for security reasons.

Reviewer 6: ```suggestion
Each `ACP99Manager` contract will be associated with one "security module" that must implement the `IACP99SecurityModule` interface and is the only contract allowed to call the `ACP99Manager` functions related to validator set changes (`initiateValidatorRegistration`, and `initiateValidatorWeightUpdate`). Everytime a validator is added/removed or a weight change is initiated, the `ACP99Manager` will in turn call the corresponding function of the "security module" (`handleValidatorRegistration` or `handleValidatorWeightChange`). We recommend that the "security module" reference an immutable `ACP99Manager` contract address for security reasons.
```

     * SubnetToL1ConversionMessage.
     * @param conversionData The Subnet conversion message data used to recompute and verify against the ConversionID.
     * @param messsageIndex The index that contains the SubnetToL1ConversionMessage Warp message containing the ConversionID to be verified against the provided {conversionData}
     */
    function initializeValidatorSet(
        ConversionData calldata conversionData,
        uint32 messsageIndex
    ) external;

    /**
     * @notice Initiate a validator registration by issuing a RegisterL1ValidatorTx Warp message
     * @param nodeID The ID of the node to add to the L1
     * @param blsPublicKey The BLS public key of the validator
     * @param registrationExpiry The time after which this message is invalid
     * @param remainingBalanceOwner The remaining balance owner of the validator
     * @param disableOwner The disable owner of the validator
     * @param weight The weight of the node on the L1
     */
    function initiateValidatorRegistration(
        bytes memory nodeID,
        bytes memory blsPublicKey,
        uint64 registrationExpiry,

>         PChainOwner memory remainingBalanceOwner,

Author: Yes makes sens

The functions `getValidation` and `getValidatorValidations` would allow to retrieve historical information about the validator set directly from state, notably each validator's performance (uptime) during past validations.
If we don't keep track of the historical information in the `ACP99Manager` contract, this information will still be available in archive nodes and offchain tools (e.g. explorers).

### Is an upgradeable architecture more appropriate?

The Teleporter team has been working on its own implementation of a `ValidatorManager` contract in the [teleporter](https://github.com/ava-labs/teleporter/blob/main/contracts/validator-manager/ValidatorManager.sol) repository that is based on abstract and upgradeable contracts.

### Should the `ACP99Manager` include a churn control mechanism?

The Teleporter team implentation of the `ValidatorManager` contract includes a churn control mechanism that prevents too much weight from being added or removed from the validator set in a short amount of time.

### How could we name “Security Modules”?

I don’t really like this name but cannot come up with anything else.

## Acknowledgments

> Special thanks to @leopaul36, @aaronbuchwald, @dhrubabasu, @minghinmatthewlam and @michaelkaplan13 for their reviews of previous versions of this ACP!

Reviewer 6: ```suggestion
Special thanks to [@leopaul36](https://github.com/leopaul36), [@aaronbuchwald](https://github.com/aaronbuchwald), [@dhrubabasu](https://github.com/dhrubabasu), [@minghinmatthewlam](https://github.com/minghinmatthewlam) and [@michaelkaplan13](https://github.com/michaelkaplan13) for their reviews of previous versions of this ACP!
```
